middleware/markdown/markdown.go

```
@@ -52,124 +42,125 @@ type Config struct {
-	StaticFiles map[string]string
+// ServeHTTP implements the http.Handler interface.
+func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+	var cfg *Config
+	for _, c := range md.Configs {
+		if middleware.Path(r.URL.Path).Matches(c.PathScope) { // not negated
+			cfg = c
+			break // or goto
+		}
+	}
+	if cfg == nil {
+		return md.Next.ServeHTTP(w, r) // exit early
+	}
 
-	// Links to all markdown pages ordered by date.
-	Links []PageLink
+	// We only deal with HEAD/GET
+	switch r.Method {
+	case http.MethodGet, http.MethodHead:
+	default:
+		return http.StatusMethodNotAllowed, nil
+	}
 
-	// Stores a directory hash to check for changes.
-	linksHash string
+	var dirents []os.FileInfo
+	var lastModTime time.Time
+	fpath := r.URL.Path
+	if idx, ok := middleware.IndexFile(md.FileSys, fpath, md.IndexFiles); ok {
+		// We're serving a directory index file, which may be a markdown
+		// file with a template.  Let's grab a list of files this directory
+		// URL points to, and pass that in to any possible template invocations,
+		// so that templates can customize the look and feel of a directory.
+		fdp, err := md.FileSys.Open(fpath)
+		if err != nil {
+			if os.IsPermission(err) {
+				return http.StatusForbidden, err
+			}
+			return http.StatusInternalServerError, err
+		}
 
-	// Directory to store static files
-	StaticDir string
+		// Grab a possible set of directory entries.  Note, we do not check
+		// for errors here (unreadable directory, for example).  It may
+		// still be useful to have a directory template file, without the
+		// directory contents being present.
+		dirents, _ = fdp.Readdir(-1)
+		for _, d := range dirents {
+			lastModTime = latest(lastModTime, d.ModTime())
+		}
 
-	// If in development mode. i.e. Actively editing markdown files.
-	Development bool
+		// Set path to found index file
+		fpath = idx
+	}
 
-	sync.RWMutex
-}
+	// If supported extension, process it
+	if _, ok := cfg.Extensions[path.Ext(fpath)]; ok {
```

Fail fast here. Invert the ```if```.

```go
if !cfg.Extensions[path.Ext(fpath)] {
  // Didn't qualify to serve as markdown; pass-thru
  return md.Next.ServeHTTP(w, r)
}

… the MarkDown stuff, one level of indentation less
```